src/main/java/net/glowstone/block/GlowBlock.java

```diff
@@ -212,48 +211,101 @@ public void setType(Material type, byte data, boolean applyPhysics) {
         setTypeIdAndData(type.getId(), data, applyPhysics);
     }
 
+    // Type ID
+    @Override
+    public int getTypeId() {
+        return getTypeId(false);
+    }
+
+    private int getTypeId(boolean noCache) {
+        if (noCache) {
+            // we need to read from storage, not from cache
+            // & 0xf ensures we get the location relative to the chunk [0 to 15]
+            return getChunk().getType(x & 0xf, z & 0xf, y);
+        }
+        if (type > -1) {
+            // cache has been initialized, just return the cached value
+            return type;
+        }
+        // set type to enable cache from stored value
+        return type = getTypeId(true);
+    }
+
     @Override
     public boolean setTypeId(int type) {
         return setTypeId(type, true);
     }
 
     @Override
     public boolean setTypeId(int type, boolean applyPhysics) {
-        return setTypeIdAndData(type, (byte) 0, applyPhysics);
+        return setTypeIdAndData(type, getData(), applyPhysics);
```

Comment:
As to keep the same functionality as Craftbukkit/Spigot, it should get reset to the default data value of the new block